### PR TITLE
[Sprint: 60] XD-3575 Add visual representation of job workflow in executions list page

### PR DESCRIPTION
**Pull request by ghillert**
* XD-3576 Add support to retrieve job details
* XD-3573 Include job-composition flag in REST endpoint

### CODE REVIEW

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/rest/BatchJobExecutionsController.java [hunk 1]
```diff
@@ -204,7 +210,22 @@ public JobExecutionInfoResource getJobExecutionInfo(@PathVariable long execution
 		final Set<String> deployedJobs = new HashSet<String>(jobLocator.getJobNames());
 		final Set<String> jobDefinitionNames = new HashSet<String>(getJobDefinitionNames());
 
-		return getJobExecutionInfoResource(jobExecution, restartableJobs, deployedJobs, jobDefinitionNames);
+		final JobExecutionInfoResource jobExecutionInfoResource = getJobExecutionInfoResource(jobExecution,
+				restartableJobs, deployedJobs, jobDefinitionNames);
+
+		for (JobExecution childJobExecution : jobService.getChildJobExecutions(jobExecution.getId())) {
+			jobExecutionInfoResource.setComposedJob(true);
+			stepExecutionDao.addStepExecutions(childJobExecution);
```
Comment (ilayaperumalg) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/rest/BatchJobExecutionsController.java]: Is this correct to set `childJobExecution` as stepExecutions here?

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/rest/BatchJobExecutionsController.java [hunk 2]
```diff
@@ -204,7 +210,22 @@ public JobExecutionInfoResource getJobExecutionInfo(@PathVariable long execution
...
+			final JobExecutionInfoResource childJobExecutionInfoResource = getJobExecutionInfoResource(
+					childJobExecution,
+					restartableJobs, deployedJobs, jobDefinitionNames);
+			boolean isCompposed = jobService.isComposedJobExecution(childJobExecution.getId());
```
Comment (ilayaperumalg) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/rest/BatchJobExecutionsController.java]: typo here: isCompposed

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dao/XdJdbcSearchableJobExecutionDao.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		setJobExecutionIncrementer(new AbstractDataFieldMaxValueIncrementer() {
+
+			@Override
+			protected long getNextKey() {
+				return 0;
+			}
+		});
+
+		final String subQuery = getQuery(
+				"not exists (select * from %PREFIX%JOB_EXECUTION_PARAMS a where key_name = 'xd_parent_execution_id' and E.JOB_EXECUTION_ID=a.JOB_EXECUTION_ID)");
```
Comment (ilayaperumalg) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dao/XdJdbcSearchableJobExecutionDao.java]: Can we move this string `xd_parent_execution_id` to a constant from where it actually originates?

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dao/XdJdbcSearchableJobExecutionDao.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		super.afterPropertiesSet();
+
+	}
+
+	/**
+	 * @return a {@link PagingQueryProvider} with a where clause to narrow the
+	 * query
+	 * @throws Exception
+	 */
+	private PagingQueryProvider getPagingQueryProvider(String fromClause, String whereClause) throws Exception {
```
Comment (ilayaperumalg) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dao/XdJdbcSearchableJobExecutionDao.java]: If the `fromClause` is always null and set inside the method, why can't we just remove this method as argument?

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/rest/BatchJobExecutionsController.java (hunk 1 above)
Comment (ghillert) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/rest/BatchJobExecutionsController.java]: A little Batch cruft there. Correct though.

**Comment by ilayaperumalg** [comment]
I have created these job definitions:

```
job create a1 --definition "timestampfile"
job create a2 --definition "timestampfile"
job deploy a1
job deploy a2
job create a3 --definition "a1 || a2" --deploy
job launch a3
```

In this case, should the job `a3` have child job executions? With the above list, I don't see any child jobexecutions for a3's job execution.
Please correct if I am missing anything here.

**Comment by ghillert** [comment]
@ilayaperumalg Are you testing in conjunction with the Admin UI PR? https://github.com/spring-projects/spring-xd-admin-ui-client/pull/10

**Comment by ghillert** [comment]
Added Shell support for composed job executions

**Comment by ilayaperumalg** [comment]
LGTM, merging.

**Comment by ilayaperumalg** [comment]
Squshed, rebased and merged as 02085a9

**Pull request closed without merging**